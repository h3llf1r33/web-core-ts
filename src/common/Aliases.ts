export type EnvironmentVariable = string;
export type HeaderName = string;