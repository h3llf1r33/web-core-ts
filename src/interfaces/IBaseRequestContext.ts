export interface IBaseRequestContext {
    headers: Record<string, string>;
    method: string;
    path: string;
    body?: any;
    queryParameters?: Record<string, string>;
    pathParameters?: Record<string, string>;
    requestId: string;
}