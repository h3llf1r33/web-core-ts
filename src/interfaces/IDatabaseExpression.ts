export interface IDatabaseExpression {
    conditions?: any;
    values?: any;
    names?: any;
    tableName?: string;
}