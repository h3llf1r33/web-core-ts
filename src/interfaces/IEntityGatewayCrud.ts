import {IEntityGatewayCreate} from "./IEntityGatewayCreate";
import {IEntityGatewayRead} from "./IEntityGatewayRead";
import {IEntityGatewayReadList} from "./IEntityGatewayReadList";
import {IEntityGatewayPatch} from "./IEntityGatewayPatch";
import {IEntityGatewayDelete} from "./IEntityGatewayDelete";
import {IEntityGateway} from "./IEntityGateway";
import {IEntityGatewayPut} from "./IEntityGatewayPut";

export interface IEntityGatewayCrud<
    CREATE_OR_UPDATE_QUERY, RESPONSE_MODEL,
    FILTER_QUERY, READ_ENTITY_ID,
    UPDATE_ENTITY_ID, DELETE_ENTITY_ID,
    DELETE_RESPONSE_MODEL
> extends IEntityGateway,
    IEntityGatewayCreate<CREATE_OR_UPDATE_QUERY, RESPONSE_MODEL>,
    IEntityGatewayRead<READ_ENTITY_ID, FILTER_QUERY, RESPONSE_MODEL>,
    IEntityGatewayReadList<FILTER_QUERY, RESPONSE_MODEL>,
    IEntityGatewayPut<UPDATE_ENTITY_ID, CREATE_OR_UPDATE_QUERY, RESPONSE_MODEL>,
    IEntityGatewayPatch<UPDATE_ENTITY_ID, CREATE_OR_UPDATE_QUERY, RESPONSE_MODEL>,
    IEntityGatewayDelete<DELETE_ENTITY_ID, DELETE_RESPONSE_MODEL> {
}