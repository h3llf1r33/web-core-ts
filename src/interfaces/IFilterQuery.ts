import {IOperator} from "./IOperator";

export interface IFilterQuery {
    field: string; // supports dot notation
    operator: IOperator;
    value: string | number | object | Array<string | number> | undefined | null;
}

export interface IPaginationQuery {
    page?: number;
    limit?: number;
    offset?: number;
    sortBy?: string;
    sortDirection?: 'asc' | 'desc';
}

export interface IPaginatedResponse<T> {
    data: T[];
    total: number;
    page: number;
    limit: number;
}

export interface IGenericFilterQuery {
    filters: IFilterQuery[];
    pagination: IPaginationQuery;
}

const operatorMap: Record<IOperator, string> = {
    '<': 'lt',
    '>': 'gt',
    '<=': 'lte',
    '>=': 'gte',
    '=': 'eq',
    '!=': 'ne',
    'in': 'in',
    'not in': 'notin',
    'like': 'like',
    'not like': 'notlike'
};

const reverseOperatorMap: Record<string, IOperator> = Object.entries(operatorMap)
    .reduce((acc, [key, value]) => ({...acc, [value]: key as IOperator}), {});

const MAX_KEY_LENGTH = 100;
const MAX_VALUE_LENGTH = 1000;
const MAX_ARRAY_LENGTH = 100;

function sanitizeKey(key: string): string {
    return key.trim();
}

function sanitizeValue(value: any): any {
    if (!Array.isArray(value) && typeof value !== 'object' && typeof value === "string") {
        return isNaN(Number(value)) ? value : Number(value)
    }
    if (typeof value === 'string') {
        const trimmed = value.slice(0, MAX_VALUE_LENGTH).trim();
        // Attempt to convert to number if the string represents a valid number
        const num = Number(trimmed);
        return !isNaN(num) && trimmed !== '' ? num : trimmed;
    }
    if (Array.isArray(value)) {
        return value
            .slice(0, MAX_ARRAY_LENGTH)
            .map(v => {
                if (typeof v === 'string') {
                    const trimmed = v.slice(0, MAX_VALUE_LENGTH).trim();
                    const num = Number(trimmed);
                    return !isNaN(num) && trimmed !== '' ? num : trimmed;
                }
                return v;
            });
    }
    if (typeof value === 'object' && value !== null) {
        return JSON.stringify(value).slice(0, MAX_VALUE_LENGTH);
    }
    return value;
}

function validateAndSanitizeFilter(filter: IFilterQuery): IFilterQuery {
    if (!filter || typeof filter !== 'object') {
        throw new Error('Invalid filter: must be an object');
    }

    if (!filter.field) {
        throw new Error('Invalid keyPath: must be a non-empty string');
    }

    const sanitizedKey = sanitizeKey(filter.field);

    if (sanitizedKey.length > MAX_KEY_LENGTH) {
        throw new Error(`KeyPath exceeds maximum length of ${MAX_KEY_LENGTH} characters`);
    }

    if (!/^[a-zA-Z0-9]+(?:\.[a-zA-Z0-9]+)*$/.test(sanitizedKey)) {
        throw new Error('Invalid keyPath format: must contain only alphanumeric characters and dots');
    }

    if (!filter.operator || !operatorMap[filter.operator]) {
        throw new Error('Invalid operator');
    }

    if (filter.value === undefined || filter.value === null) {
        throw new Error('value cannot be null or undefined');
    }

    const sanitizedValue = sanitizeValue(filter.value);

    if ((filter.operator === 'in' || filter.operator === 'not in') && !Array.isArray(filter.value)) {
        throw new Error('value must be an array for "in" and "not in" operators');
    }

    return {
        field: sanitizedKey,
        operator: filter.operator,
        value: sanitizedValue
    };
}

/**
 * Serializes IGenericFilterQuery into URL query parameters.
 * @param query The generic filter query.
 * @returns A string representing the serialized query parameters.
 */
export function serializeGenericFilterQuery(query: IGenericFilterQuery): string {
    const params: Record<string, string> = {};
    query.filters = query.filters.map((data) => ({
        field: data.field,
        operator: data.operator,
        value: sanitizeValue(data.value)
    }));

    if (query.filters && query.filters.length > 0) {
        params.filters = encodeURIComponent(JSON.stringify(query.filters));
    }

    if (query.pagination) {
        params.pagination = encodeURIComponent(JSON.stringify(query.pagination));
    }

    const queryString = Object.entries(params)
        .map(([key, value]) => `${key}=${value}`)
        .join('&');

    return queryString ? `?${queryString}` : '';
}


/**
 * Deserializes URL query parameters into IGenericFilterQuery.
 * @param queryString The URL query string.
 * @returns The deserialized generic filter query.
 */
export function deserializeGenericFilterQuery(queryString: string): IGenericFilterQuery {
    const params = new URLSearchParams(queryString);
    const filtersParam = params.get('filters');
    const paginationParam = params.get('pagination');

    let filters: IFilterQuery[] = [];
    let pagination: IPaginationQuery = {};

    try {
        if (filtersParam) {
            const parsedFilters: IFilterQuery[] = JSON.parse(decodeURIComponent(filtersParam));
            filters = parsedFilters.map(filter => validateAndSanitizeFilter(filter));
        }
    } catch (error) {
        console.error('Error parsing filters:', error);
    }

    try {
        if (paginationParam) {
            pagination = JSON.parse(decodeURIComponent(paginationParam));
        }
    } catch (error) {
        console.error('Error parsing pagination:', error);
    }

    return {filters, pagination};
}

export function convertFilterToQueryParams(filter: IFilterQuery | IFilterQuery[]): string {
    if (!filter) return '';

    const filters = Array.isArray(filter) ? filter : [filter];
    if (!filters.length) return '';

    if (filters.length > MAX_ARRAY_LENGTH) {
        throw new Error(`Too many filters: maximum allowed is ${MAX_ARRAY_LENGTH}`);
    }

    const params = filters.map(filter => {
        const sanitizedFilter = validateAndSanitizeFilter(filter);
        let processedValue = sanitizedFilter.value;

        if (Array.isArray(processedValue)) {
            processedValue = processedValue.join(',');
        } else if (typeof processedValue === 'object') {
            processedValue = JSON.stringify(processedValue);
        }

        const encodedValue = encodeURIComponent(String(processedValue));
        return `${sanitizedFilter.field}[${operatorMap[sanitizedFilter.operator]}]=${encodedValue}`;
    });

    return params.length ? `?${params.join('&')}` : '';
}

export function parseQueryParamsToFilter(queryString: string): IFilterQuery[] {
    if (!queryString) return [];

    const query = queryString.startsWith('?') ? queryString.slice(1) : queryString;
    if (!query) return [];

    const params = query.split('&');
    if (params.length > MAX_ARRAY_LENGTH) {
        throw new Error(`Too many parameters: maximum allowed is ${MAX_ARRAY_LENGTH}`);
    }

    return params.map(param => {
        const match = param.match(/^([^[\]]+)\[([^\]]+)]=(.+)$/);
        if (!match) {
            throw new Error(`Invalid parameter format: ${param}`);
        }

        const [, keyPath, operator, encodedValue] = match;
        const mappedOperator = reverseOperatorMap[operator];

        if (!mappedOperator) {
            throw new Error(`Invalid operator: ${operator}`);
        }

        const decodedValue = decodeURIComponent(encodedValue);
        let parsedValue: any = decodedValue;

        if (mappedOperator === 'in' || mappedOperator === 'not in') {
            parsedValue = decodedValue.split(',')
                .map(v => {
                    const num = Number(v);
                    return !isNaN(num) ? num : v.trim();
                });
        } else {
            const numberValue = Number(decodedValue);
            if (!isNaN(numberValue) && decodedValue !== '') {
                parsedValue = numberValue;
            } else if (decodedValue.startsWith('{') || decodedValue.startsWith('[')) {
                try {
                    parsedValue = JSON.parse(decodedValue);
                } catch {
                    parsedValue = decodedValue;
                }
            }
        }

        return validateAndSanitizeFilter({
            field: keyPath,
            operator: mappedOperator,
            value: parsedValue
        });
    });
}
