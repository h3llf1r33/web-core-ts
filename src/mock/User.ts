export interface IUser {
    id: string;
    name: string;
}

export const mockUsers = (quantity: number): IUser[] => {
    return Array.from(Array(quantity).keys()).map((_, index): IUser => {
        const id = index + 1;
        return {
            id: String(id),
            name: `John Doe ${id}`
        }
    })
}

export const mockUser: IUser = {
    id: "1",
    name: "John Doe"
}