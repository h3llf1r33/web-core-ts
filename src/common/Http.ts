import {Observable} from "rxjs";
import {IHttpHeaders} from "../interfaces/IHttpHeaders";

export type HttpClientMiddleware<T> = ((config: T) => Observable<T>);
export type HttpMethodType = 'GET' | 'POST' | 'PUT' | 'PATCH' | 'DELETE';

export type HttpClientRequestOptions =
    Omit<Axios.AxiosXHRConfig<Record<string, any>>, 'url' | 'method' | 'data' | 'headers'>
    & {
    headers?: IHttpHeaders;
};