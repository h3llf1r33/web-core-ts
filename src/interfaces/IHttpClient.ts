import {Observable} from "rxjs";
import {IGenericFilterQuery} from "./IFilterQuery";
import {HttpClientRequestOptions, HttpMethodType} from "../common/Http";

export interface IHttpClient {
    baseUrl: string

    get<T>(path: string, config?: HttpClientRequestOptions, filterQuery?: IGenericFilterQuery): Observable<T>;

    getRequest<T>(path: string, config?: HttpClientRequestOptions, filterQuery?: IGenericFilterQuery): Observable<Axios.AxiosXHR<T>>;

    post<T>(path: string, body?: Record<string, any>, config?: HttpClientRequestOptions): Observable<T>;

    postRequest<T>(path: string, body?: Record<string, any>, config?: HttpClientRequestOptions): Observable<Axios.AxiosXHR<T>>;

    put<T>(path: string, body?: Record<string, any>, config?: HttpClientRequestOptions): Observable<T>;

    putRequest<T>(path: string, body?: Record<string, any>, config?: HttpClientRequestOptions): Observable<Axios.AxiosXHR<T>>;

    patch<T>(path: string, body?: Record<string, any>, config?: HttpClientRequestOptions): Observable<T>;

    patchRequest<T>(path: string, body?: Record<string, any>, config?: HttpClientRequestOptions): Observable<Axios.AxiosXHR<T>>;

    delete<T>(path: string, config?: HttpClientRequestOptions): Observable<T>;

    deleteRequest<T>(path: string, config?: HttpClientRequestOptions): Observable<Axios.AxiosXHR<T>>;

    request<T, R extends boolean = false>(
        method: HttpMethodType,
        path: string,
        options: {
            config?: HttpClientRequestOptions;
            body?: Record<string, any>;
            returnFullResponse?: R;
        },
        filterQuery?: IGenericFilterQuery): Observable<R extends true ? Axios.AxiosXHR<T> : T>;
}