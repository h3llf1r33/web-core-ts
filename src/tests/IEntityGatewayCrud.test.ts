import {Observable, of} from "rxjs";
import {IUser, mockUser, mockUsers} from "../mock/User";
import axios from "axios";
import AxiosMockAdapter from "axios-mock-adapter"
import {IEntityGatewayCrud} from "../interfaces/IEntityGatewayCrud";
import {IGenericFilterQuery, IPaginatedResponse} from "../interfaces/IFilterQuery";
import {IHttpClient} from "../interfaces/IHttpClient";
import {HttpClientRequestOptions} from "../common/Http";
import {HttpClientAxios} from "@denis_bruns/http-client-axios";


const mock = new AxiosMockAdapter(axios);

let BASE_URL = '';

mock.onGet("/users").reply(200, mockUsers(100));
mock.onGet(/\/user\/\d+/).reply(200, mockUser);
mock.onPut(/\/user\/\d+/).reply(200, mockUser);
mock.onPatch(/\/user\/\d+/).reply(200, mockUser);
mock.onPost("/user").reply(201, mockUser);
mock.onDelete(/\/user\/\d+/).reply(200, true);

class TestCrudGateway implements IEntityGatewayCrud<
    IUser, IUser, IGenericFilterQuery,
    string, string, string, boolean
> {
    constructor(private httpClient: IHttpClient) {
    }

    create(query: Partial<IUser>): Observable<IUser> {
        return this.httpClient.post<IUser>("/user", query)
    }

    createAJAX(query: Partial<IUser>): Observable<Axios.AxiosXHR<IUser>> {
        return this.httpClient.postRequest<IUser>("/user", query)
    }

    read(query?: string, filterQuery?: IGenericFilterQuery): Observable<IUser> {
        return this.httpClient.get<IUser>(`/user/${query}`, {}, filterQuery)
    }

    readAJAX(query?: string, filterQuery?: IGenericFilterQuery): Observable<Axios.AxiosXHR<IUser>> {
        return this.httpClient.getRequest<IUser>(`/user/${query}`, {}, filterQuery)
    }

    readList(filterQuery?: IGenericFilterQuery): Observable<IPaginatedResponse<IUser>> {
        return this.httpClient.get<IPaginatedResponse<IUser>>("/users", {}, filterQuery)
    }

    readListAJAX(filterQuery?: IGenericFilterQuery): Observable<Axios.AxiosXHR<IUser[]>> {
        return this.httpClient.getRequest<IUser[]>("/users", {}, filterQuery)
    }

    updateEntity(entityId: string, query: Partial<IUser>): Observable<IUser> {
        return this.httpClient.patch<IUser>(`/user/${entityId}`, query)
    }

    updateEntityAJAX(entityId: string, query: Partial<IUser>): Observable<Axios.AxiosXHR<IUser>> {
        return this.httpClient.patchRequest<IUser>(`/user/${entityId}`, query)
    }

    replaceEntity(entityId: string, query: Partial<IUser>): Observable<IUser> {
        return this.httpClient.put<IUser>(`/user/${entityId}`, query)
    }

    replaceEntityAJAX(entityId: string, query: Partial<IUser>): Observable<Axios.AxiosXHR<IUser>> {
        return this.httpClient.putRequest<IUser>(`/user/${entityId}`, query)
    }

    delete(entityId: string): Observable<boolean> {
        return this.httpClient.delete<boolean>(`/user/${entityId}`)
    }

    deleteAJAX(entityId: string): Observable<Axios.AxiosXHR<boolean>> {
        return this.httpClient.deleteRequest<boolean>(`/user/${entityId}`)
    }
}

describe('', () => {
    test('should be able to fetch data over each method', async () => {
        const crudEntityGateway = new TestCrudGateway(new HttpClientAxios(""));

        const readData = await crudEntityGateway.read("1", {
            filters: [
                {
                    field: 'name',
                    operator: "=",
                    value: 'hello world'
                }
            ], pagination: {}
        }).toPromise();
        expect(readData).toMatchSnapshot();

        const readListData = await crudEntityGateway.readList().toPromise();
        expect(readListData).toMatchSnapshot();

        const createData = await crudEntityGateway.create(mockUser).toPromise();
        expect(createData).toMatchSnapshot();

        const patchData = await crudEntityGateway.updateEntity("1", mockUser).toPromise();
        expect(patchData).toMatchSnapshot();

        const putData = await crudEntityGateway.replaceEntity("1", mockUser).toPromise();
        expect(putData).toMatchSnapshot();

        const deleteData = await crudEntityGateway.delete("1").toPromise();
        expect(deleteData).toMatchSnapshot();
    });

    test('should have modified Content-Type over middleware for each method', async () => {
        const crudEntityGateway = new TestCrudGateway(new HttpClientAxios("", [
            (config:HttpClientRequestOptions) => {
                if (!config.headers) config.headers = {}
                config.headers['Content-Type'] = 'application/fantasy'
                return of(config)
            }
        ]));

        const readData = await crudEntityGateway.readAJAX("1").toPromise();
        expect(readData?.config?.headers?.['Content-Type']).toMatchSnapshot();

        const readListData = await crudEntityGateway.readListAJAX().toPromise();
        expect(readListData?.config?.headers?.['Content-Type']).toMatchSnapshot();

        const createData = await crudEntityGateway.createAJAX(mockUser).toPromise();
        expect(createData?.config?.headers?.['Content-Type']).toMatchSnapshot();

        const patchData = await crudEntityGateway.updateEntityAJAX("1", mockUser).toPromise();
        expect(patchData?.config?.headers?.['Content-Type']).toMatchSnapshot();

        const putData = await crudEntityGateway.replaceEntityAJAX("1", mockUser).toPromise();
        expect(putData?.config?.headers?.['Content-Type']).toMatchSnapshot();

        const deleteData = await crudEntityGateway.deleteAJAX("1").toPromise();
        expect(deleteData?.config?.headers?.['Content-Type']).toMatchSnapshot();
    });
});
  