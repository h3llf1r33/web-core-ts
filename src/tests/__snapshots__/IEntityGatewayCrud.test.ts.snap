// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` should be able to fetch data over each method 1`] = `
{
  "id": "1",
  "name": "John Doe",
}
`;

exports[` should be able to fetch data over each method 2`] = `
[
  {
    "id": "1",
    "name": "John Doe 1",
  },
  {
    "id": "2",
    "name": "John Doe 2",
  },
  {
    "id": "3",
    "name": "John Doe 3",
  },
  {
    "id": "4",
    "name": "John Doe 4",
  },
  {
    "id": "5",
    "name": "John Doe 5",
  },
  {
    "id": "6",
    "name": "John Doe 6",
  },
  {
    "id": "7",
    "name": "John Doe 7",
  },
  {
    "id": "8",
    "name": "John Doe 8",
  },
  {
    "id": "9",
    "name": "John Doe 9",
  },
  {
    "id": "10",
    "name": "John Doe 10",
  },
  {
    "id": "11",
    "name": "John Doe 11",
  },
  {
    "id": "12",
    "name": "John Doe 12",
  },
  {
    "id": "13",
    "name": "John Doe 13",
  },
  {
    "id": "14",
    "name": "John Doe 14",
  },
  {
    "id": "15",
    "name": "John Doe 15",
  },
  {
    "id": "16",
    "name": "John Doe 16",
  },
  {
    "id": "17",
    "name": "John Doe 17",
  },
  {
    "id": "18",
    "name": "John Doe 18",
  },
  {
    "id": "19",
    "name": "John Doe 19",
  },
  {
    "id": "20",
    "name": "John Doe 20",
  },
  {
    "id": "21",
    "name": "John Doe 21",
  },
  {
    "id": "22",
    "name": "John Doe 22",
  },
  {
    "id": "23",
    "name": "John Doe 23",
  },
  {
    "id": "24",
    "name": "John Doe 24",
  },
  {
    "id": "25",
    "name": "John Doe 25",
  },
  {
    "id": "26",
    "name": "John Doe 26",
  },
  {
    "id": "27",
    "name": "John Doe 27",
  },
  {
    "id": "28",
    "name": "John Doe 28",
  },
  {
    "id": "29",
    "name": "John Doe 29",
  },
  {
    "id": "30",
    "name": "John Doe 30",
  },
  {
    "id": "31",
    "name": "John Doe 31",
  },
  {
    "id": "32",
    "name": "John Doe 32",
  },
  {
    "id": "33",
    "name": "John Doe 33",
  },
  {
    "id": "34",
    "name": "John Doe 34",
  },
  {
    "id": "35",
    "name": "John Doe 35",
  },
  {
    "id": "36",
    "name": "John Doe 36",
  },
  {
    "id": "37",
    "name": "John Doe 37",
  },
  {
    "id": "38",
    "name": "John Doe 38",
  },
  {
    "id": "39",
    "name": "John Doe 39",
  },
  {
    "id": "40",
    "name": "John Doe 40",
  },
  {
    "id": "41",
    "name": "John Doe 41",
  },
  {
    "id": "42",
    "name": "John Doe 42",
  },
  {
    "id": "43",
    "name": "John Doe 43",
  },
  {
    "id": "44",
    "name": "John Doe 44",
  },
  {
    "id": "45",
    "name": "John Doe 45",
  },
  {
    "id": "46",
    "name": "John Doe 46",
  },
  {
    "id": "47",
    "name": "John Doe 47",
  },
  {
    "id": "48",
    "name": "John Doe 48",
  },
  {
    "id": "49",
    "name": "John Doe 49",
  },
  {
    "id": "50",
    "name": "John Doe 50",
  },
  {
    "id": "51",
    "name": "John Doe 51",
  },
  {
    "id": "52",
    "name": "John Doe 52",
  },
  {
    "id": "53",
    "name": "John Doe 53",
  },
  {
    "id": "54",
    "name": "John Doe 54",
  },
  {
    "id": "55",
    "name": "John Doe 55",
  },
  {
    "id": "56",
    "name": "John Doe 56",
  },
  {
    "id": "57",
    "name": "John Doe 57",
  },
  {
    "id": "58",
    "name": "John Doe 58",
  },
  {
    "id": "59",
    "name": "John Doe 59",
  },
  {
    "id": "60",
    "name": "John Doe 60",
  },
  {
    "id": "61",
    "name": "John Doe 61",
  },
  {
    "id": "62",
    "name": "John Doe 62",
  },
  {
    "id": "63",
    "name": "John Doe 63",
  },
  {
    "id": "64",
    "name": "John Doe 64",
  },
  {
    "id": "65",
    "name": "John Doe 65",
  },
  {
    "id": "66",
    "name": "John Doe 66",
  },
  {
    "id": "67",
    "name": "John Doe 67",
  },
  {
    "id": "68",
    "name": "John Doe 68",
  },
  {
    "id": "69",
    "name": "John Doe 69",
  },
  {
    "id": "70",
    "name": "John Doe 70",
  },
  {
    "id": "71",
    "name": "John Doe 71",
  },
  {
    "id": "72",
    "name": "John Doe 72",
  },
  {
    "id": "73",
    "name": "John Doe 73",
  },
  {
    "id": "74",
    "name": "John Doe 74",
  },
  {
    "id": "75",
    "name": "John Doe 75",
  },
  {
    "id": "76",
    "name": "John Doe 76",
  },
  {
    "id": "77",
    "name": "John Doe 77",
  },
  {
    "id": "78",
    "name": "John Doe 78",
  },
  {
    "id": "79",
    "name": "John Doe 79",
  },
  {
    "id": "80",
    "name": "John Doe 80",
  },
  {
    "id": "81",
    "name": "John Doe 81",
  },
  {
    "id": "82",
    "name": "John Doe 82",
  },
  {
    "id": "83",
    "name": "John Doe 83",
  },
  {
    "id": "84",
    "name": "John Doe 84",
  },
  {
    "id": "85",
    "name": "John Doe 85",
  },
  {
    "id": "86",
    "name": "John Doe 86",
  },
  {
    "id": "87",
    "name": "John Doe 87",
  },
  {
    "id": "88",
    "name": "John Doe 88",
  },
  {
    "id": "89",
    "name": "John Doe 89",
  },
  {
    "id": "90",
    "name": "John Doe 90",
  },
  {
    "id": "91",
    "name": "John Doe 91",
  },
  {
    "id": "92",
    "name": "John Doe 92",
  },
  {
    "id": "93",
    "name": "John Doe 93",
  },
  {
    "id": "94",
    "name": "John Doe 94",
  },
  {
    "id": "95",
    "name": "John Doe 95",
  },
  {
    "id": "96",
    "name": "John Doe 96",
  },
  {
    "id": "97",
    "name": "John Doe 97",
  },
  {
    "id": "98",
    "name": "John Doe 98",
  },
  {
    "id": "99",
    "name": "John Doe 99",
  },
  {
    "id": "100",
    "name": "John Doe 100",
  },
]
`;

exports[` should be able to fetch data over each method 3`] = `
{
  "id": "1",
  "name": "John Doe",
}
`;

exports[` should be able to fetch data over each method 4`] = `
{
  "id": "1",
  "name": "John Doe",
}
`;

exports[` should be able to fetch data over each method 5`] = `
{
  "id": "1",
  "name": "John Doe",
}
`;

exports[` should be able to fetch data over each method 6`] = `true`;

exports[` should have modified Content-Type over middleware for each method 1`] = `"application/fantasy"`;

exports[` should have modified Content-Type over middleware for each method 2`] = `"application/fantasy"`;

exports[` should have modified Content-Type over middleware for each method 3`] = `"application/fantasy"`;

exports[` should have modified Content-Type over middleware for each method 4`] = `"application/fantasy"`;

exports[` should have modified Content-Type over middleware for each method 5`] = `"application/fantasy"`;

exports[` should have modified Content-Type over middleware for each method 6`] = `"application/fantasy"`;
